apis/v1/gatewayclass_types_overrides.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import "encoding/json"
+
+// Below code handles the experimental field breaking change introduced in
+// https://github.com/kubernetes-sigs/gateway-api/pull/3200/.
+// We are overriding the UnmarshalJSON function to be able to handle cases where
+// users had the old version of the GatewayClass CRD applied with SupportedFeatures
+// as a list of strings and not list of objects.
+// See https://github.com/istio/istio/issues/53846 for more information.
+
+func (s *SupportedFeature) UnmarshalJSON(data []byte) error {
+	var oldSupportedFeature OldSupportedFeature
+	if err := json.Unmarshal(data, &oldSupportedFeature); err == nil {
+		s.Name = FeatureName(oldSupportedFeature)
+		return nil
+	}
+
+	var si SupportedFeatureInternal
+	if err := json.Unmarshal(data, &si); err != nil {
+		return err
+	}
+	s.Name = si.Name
+	return nil
+}
+
+// This is solely for the purpose of ensuring backward compatibility and
+// SHOULD NOT be used elsewhere.
+type SupportedFeatureInternal struct {
```

Comment:
```suggestion
type supportedFeatureInternal struct {
```

nit, might want to keep it private

Reply:
Same for OldSupportedFeature

Reply:
Why do we need such a struct? Can't we just use the already existing 

https://github.com/kubernetes-sigs/gateway-api/blob/3f8a9171785096d877ed887ea4b58283cdf58bdd/apis/v1/gatewayclass_types.go#L286-L288
?

Reply:
Then it would be recursive I think. We could embed that one though I suspect (or just make it a `type supportedFeatureInternal SupportedFeature` perhaps

Reply:
done, thanks